### ISSUE TITLE
How to use custom-built Torchtext vocabulary with the HuggingFace TransfoXLLMHeadModel?

### ISSUE THREAD

**Issue by h56cho** [pytorch/text#629]
Hello,

I am trying to use my custom built vocabulary which I defined using Torchtext functions with the HuggingFace TransfoXLLMHeadModel, and I am having some troubles with it.
I defined my text field as below:
```js

# Import packages 
import torch
import torch.nn as nn
import torch.nn.functional as F
from transformers import TransfoXLConfig, TransfoXLTokenizer, TransfoXLLMHeadModel
from transformers import AdamW, WarmupLinearSchedule
import spacy
import torchtext
from torchtext.data.utils import get_tokenizer
from torchtext.data import Field, BPTTIterator, TabularDataset 
import tensorflow as tf
#import lineflow as lf
#import lineflow.datasets as lfds
import math
import random
import numpy as np
import pandas as pd 
import time

# define tokenizer
en = spacy.load('en')

def Sp_Tokenizer(text): 
    return [tok.text for tok in en.tokenizer(text)]

# define the English text field
TEXT = Field(tokenize = Sp_Tokenizer,
             init_token='< sos >',
             eos_token='< eos >',
             unk_token='< unk >',
             tokenizer_language='en',
             lower=True)

# load WikiText-2 dataset and split it into train and test set
train_Wiki2, val_Wiki2, test_Wiki2 = torchtext.datasets.WikiText2.splits(TEXT)
train_Wiki103, val_Wiki103, test_Wiki103 = torchtext.datasets.WikiText103.splits(TEXT)
train_Penn, val_Penn, test_Penn = torchtext.datasets.PennTreebank.splits(TEXT)

# build custom vocabulary based on the field that we just defined.
TEXT.build_vocab(train_Wiki2, val_Wiki2, test_Wiki2, 
                 train_Wiki103, val_Wiki103, test_Wiki103,
                 train_Penn, val_Penn, test_Penn)
```
and then I defined the HuggingFace transformer's configuration as below:
```js

# set hyperparameter ntokens
ntokens = len(TEXT.vocab.stoi)

# define transformer-XL configuration.
transfoXLconfig = TransfoXLConfig(vocab_size_or_config_json_file = ntokens,
                                  cutoffs = [20000, 40000, 200000], 
                                  d_model = 64, 
                                  d_embed = 64, 
                                  n_head = 16, 
                                  d_head = 64,
                                  n_layer = 5,
                                  attn_type = 0,
                                  dropout = 0.1, 
                                  output_hidden_states = True,
                                  output_attentions = True)

# define the transformer-XL model based on the specified configuration.
model = TransfoXLLMHeadModel(transfoXLconfig)

# add new tokens to the embeddings of our model
model.resize_token_embeddings(ntokens)
```
and then I want to somehow specify that I want to use my `TEXT.vocab` that I defined earlier via Torchtext for my vocabulary along with the TransfoXLLMHeadModel, but I am not sure how to do this. Can someone help me on this? Thank you!

**Comment by zhangguanheng66** [comment]
You mean apply HF tokenizer, instead of spacy? Since you have the train/valid/test data, you should be able to send them to HF model for training.

BTW, the vocab is generated from train/valid/test data. Why not just train data?

**Comment by h56cho** [comment]
Hello,

Thank you for your reply.
The problem I am facing is I don't know how to send my train/valid/test data to a HF model for training. Don't I need to write a separate train function for training the model?

Below is my BPTTiterator and the train function.
Do you see any error in my train function? Thank you,

```js
train_iter, val_iter, test_iter = BPTTIterator.splits(
            (train_Wiki2, val_Wiki2, test_Wiki2),
            batch_size = batch_size,
            bptt_len= bptt,
            sort_key=lambda x: len(x.text),
            sort_within_batch = True,
            shuffle = False,
            device= device,
            repeat=False)

train = next(iter(train_iter))
val = next(iter(train_iter))
test = next(iter(test_iter))

model.train()


# define the train function
def train(model, train, bptt, criterion, optimizer, scheduler, ntokens, log_interval):

    # initialize total_loss to 0
    total_loss = 0
    # measure the computation time
    start_time = time.time()
    # number of tokens in the vocabulary
    ntokens = ntokens 

    for i in range(train.text.size()[1]):
        
        batch = i
        input_ids, targets = train.text[:,i], train.target[:,i]
        input_ids = torch.tensor(input_ids.tolist()).unsqueeze(0) 
        targets = torch.tensor(targets.tolist()).unsqueeze(0)
        optimizer.zero_grad() # set the gradient back to 0 (necessary step)
        output = model(input_ids, labels = targets)
        loss = output[0]

        loss.backward()
        
        torch.nn.utils.clip_grad_norm_(model.parameters(), 0.5)
        optimizer.step() 
```

NOTE: I am especially interested in knowing whether the part of my train function below is correct:

```
output = model(input_ids, labels = targets)
loss = output[0]
loss.backward()
```
What I am intending to do here is to train my custom Transformer-XL model by calculating the loss and propagate it backwards. 

Also, it seems that this HuggingFace integration is not allowing users to specify the type of loss that's being calculated --- in the `loss = output[0]`, what type of loss does this specifically refer to? Thank you,

**Comment by h56cho** [comment]
Could you please provide me with some example code on how to send the datasets generated by `train_Wiki2, val_Wiki2, test_Wiki2 = torchtext.datasets.WikiText2.splits(TEXT)` to a custom HF model for training?

**Comment by zhangguanheng66** [comment]
[Here](https://pytorch.org/tutorials/beginner/torchtext_translation_tutorial.html) is an example.

Please note that the datasets in torchtext are subject to change at this moment. Dataset with new pattern (more compatible with `torch.utils.data`) could be found [here](https://pytorch.org/tutorials/beginner/text_sentiment_ngrams_tutorial.html).

**Comment by mttk** [comment]
Hi,

your `train` function is more or less correct. You can iterate over the training data without indexing like this:
```python
for idx, (inputs, targets) in enumerate(train):
  ...
```

The return value from the `model()` call can be seen if you check the class implementation in the `transformers` library here: https://github.com/huggingface/transformers/blob/master/transformers/modeling_transfo_xl.py#L780

The return values are always outputs of a (variant) of a softmax function, and a concrete loss isn't implemented. This can also be seen from this issue, indicating that transformerXL can't yet be trained from scratch: https://github.com/huggingface/transformers/issues/1639 

If the model from the huggingface repo is implemented for training/fine-tuninng, it would also return a concrete loss and your training function should work fine. I'd suggest you either wait for the model to be implemented completely, use a different model that is implemented completely, or implement the remainder yourself.

**Comment by h56cho** [comment]
Hello, if HuggingFace transformerXL model can't yet be trained from scratch and loss is not returned as the result, why is the HuggingFace Transformer XL documentation indicate that the loss can be returned with appropriate labels(https://huggingface.co/transformers/model_doc/transformerxl.html#transfoxllmheadmodel)?

**Comment by mttk** [comment]
I believe it is better to ask that question at the trasnformers repository.

**Comment by h56cho** [comment]
Hello,
Thank you again for your reply. I just found out that it was an error in the documentation, and like you mentioned, the TransfoXLLMHead model cannot be trained from scratch just yet.

Because I am new to python and NLP in general, can I confirm with you that, unlike TransfoXLLMHead model, the HuggingFace TransfoXLModel is trainable from scratch. From the TransfoXLModel code, it appears that the concrete last hidden state can be returned via TransfoXLModel, and perhaps I could train it from scratch by executing the code below:

```js
criterion = nn.CrossEntropyLoss()
loss = criterion(output.view(-1, ntokens), targets)
```

 instead of `loss = output[0]` like I did for the TransfoXLLMHeadsModel.

Am I correct on this?

Thank you again,

**Comment by mttk** [comment]
The `TransfoXLModel` is merely en encoder, and is not task-specific. Given a sequence of inputs it will produce a sequence of hidden states representing the "contextualized" sequence. In order to train it, you need to apply it to a concrete task (done by adding a "prediction head", hence `TransfoXLLMHead`, which does language modelling) So, you would need to make your changes to the `TransfoXLLMHead` model.

I haven't yet invested time to read and understand the transformer-XL training process, however I can see from the code that it uses modifications of the softmax function (sampled or projected adaptive). I am not sure if in that case it is enough to simply apply the `CrossEntropyLoss`, and I am more inclined to believe that it is not enough. Apart from directing you to read the [paper](https://arxiv.org/abs/1901.02860) and try to determine yourself, there isn't much I can do to help.
